File: leetcode/lc-us/cpp/64.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int minPathSum(vector<vector<int>> &grid) {
        vector<vector<int>> dp(grid);
        for (int i = 1; i < dp.size(); i++) dp[i][0] += dp[i - 1][0];
        for (int j = 1; j < dp[0].size(); j++) dp[0][j] += dp[0][j - 1];
        for (int i = 1; i < dp.size(); i++) {
            for (int j = 1; j < dp[0].size(); j++) {
                dp[i][j] += min(dp[i - 1][j], dp[i][j - 1]);
            }
        }
        return dp.back().back();
    }
};

File: leetcode/lc-us/rust/64.rs
use std::cmp::min;

impl Solution {
    pub fn min_path_sum(grid: Vec<Vec<i32>>) -> i32 {
        let n = grid.len();
        let m = grid[0].len();
        let mut f: Vec<Vec<i32>> = vec![vec![0; m]; n];
        f[0][0] = grid[0][0];
        for j in 1..m {
            f[0][j] = f[0][j - 1] + grid[0][j];
        }
        for i in 1..n {
            f[i][0] = f[i - 1][0] + grid[i][0];
        }
        for i in 1..n {
            for j in 1..m {
                f[i][j] = min(f[i - 1][j], f[i][j - 1]) + grid[i][j];
            }
        }
        return f[n - 1][m - 1];
    }
}